man/pages/fish.man
NAME
	fish - the card game "Go Fish", played against the computer

SYNOPSIS
	fish [-p]

DESCRIPTION
	Fish is a CP/M port of the BSD Unix game of the same name.

OPTIONS
	-p	`pro' mode - make the computer's game stronger.

PLAYING THE GAME
	This section just quotes the instructions from the original
	Unix version, which apply equally to this CP/M port:

	`This is the traditional children's card game "Go Fish". We
	each get seven cards, and the rest of the deck is kept to be
	drawn from later. The object of the game is to collect
	"books", or all of the cards of a single value. For example,
	getting four 2's would give you a "book of 2's".

	We take turns asking each other for cards, but you can't ask
	me for a card value if you don't have one of them in your
	hand! If I have any cards of the value you ask for, I have to
	give them to you. As long as I have one of the cards you ask
	for, you get to keep asking. If you ask me for a card of which
	I don't have any, then I'll tell you to "Go Fish!" This means
	that you draw a card from the deck. If you draw the card you
	asked me for, you get to keep asking me for cards. If not,
	it's my turn and I ask you for a card.

	Sometimes you get to ask first, sometimes I do. I'll tell you
	when it's your turn to move, I'll draw cards from the deck for
	you, and I'll tell you what you have in your hand. (Don't
	worry, I don't look at your hand when I'm trying to decide
	what card to ask for, honest!)

	Your input can be a card name ("A", "2", "3", "4", "5", "6",
	"7", "8", "9", "10", "J", "Q" or "K") or the letter "p", or
	"quit". The letter "p" makes my game much smarter, and the
	line "quit" stops the game. Just hitting the carriage return
	key displays how many cards I have in my hand, how many are
	left in the deck, and which books I've gotten.

	Normally, the game stops when one of us runs out of cards, and
	the winner is whoever has the most books!

	Good luck!'

AUTHOR
	Muffy Barkocy


NEWS
* Changes between versions   -*- mode:indented-text; mode:outline-minor -*-

This file (NEWS) documents the significant changes between all
versions of ZCN since 1.2, with the newest changes first.


* Changes in ZCN 1.3

Some important bug fixes:

(The first three were also fixed in zcn12fix.zip, so they probably
won't be `new' fixes to most people)

man was totally broken; it would mysteriously die without giving any
reason, which would lead to subtle crashes later on. This was due to
the most stupid thinko I think I've ever perpetrated. :-(

A less stupid, but still significant, thinko meant that only 1992 and
1996 were treated as leap years. (2000, 2004, etc. weren't.) ZCN now
has no Y2K problems (it's ok until 2099). But be sure to set the time
and date before 2004 so it has a chance to fix the old problem. :-)

Previously, if you asked rrxfer for a list of the ROM files and there
were none, it crashed (ouch!).

In the context-restore routine, the stack pointer was pointing at the
TPA at a Bad Time when restoring things, meaning that memory in the
B1F8h-B1FDh range (6 bytes) got blasted. Not the sort of bug you'd
usually notice easily, but it happened to break wade.

The context-restore rewrote 0038h-003Ah despite not needing to, which
could probably have broken wade in some (relatively unusual)
circumstances.

spellwd used to crash when run without args, when spellwd wasn't the
last command you'd run.

zcnpaint's eraser tool previously drew horizontal/vertical lines when
used right at the top/left of the screen.

And, incredibly, there's some non-bugfix stuff too... :-)

The `kernel' now assembles into two versions - the original zcn.bin
for NC100s, and a new zcn200.bin for NC200s. The NC200 version still
has some rough edges, but supports most of the hardware except for the
disk drive. Thanks to Ian Miller for helping a *lot* with testing the
NC200 support (as I don't actually have an NC200).

Added `serial' command, which allows you to disable/re-enable the
serial line driver. It's enabled by default (just as in previous
versions of ZCN), but disabling it should prolong battery life. Thanks
to Emmanuel Roussin for bringing this up.

`bigv' views a single file in much the same way as with `man's viewer.
It uses a large, proportionally-spaced font - the intent is that bigv
be comfortable to read from, especially for large texts such as online
books (`etexts'). Thanks to Steven Flintham for the inspiration behind
this.

Some minor speed optimisations in a BDOS hotspot (the crawling horror
that is `matchfcb' :-)).

`keyb' changes the keyboard map. Currently this only supports the
default UK map, and Dvorak for the UK layout. Thanks to Ian Miller for
suggesting Dvorak support.

`lfconv' converts Unix-format text files to CP/M (CR/LF) format.

VDE has replaced ZDE as the text editor included with ZCN, mainly
because it's the more widely-used version (ZDE is a VDE derivative),
and it comes with full documentation. But don't worry, your old copy
of ZDE will obviously still work if you'd prefer to stick with that.

Removed nswp from distribution.


* Changes in ZCN 1.2

Decent online help. This comes in the shape of `man', a Unix-man-like
program, and *many* man pages for it, documenting every ZCN command
and most third-party ones bundled with ZCN, taking roughly 55k for the
lot. The `help' program has been removed, as man makes it obsolete.

`zap', a graphical program-launcher.

`spell', an interactive spellchecker (which uses the ROM), a bit like
a cross between `ispell' and the ROM wp's spellchecker.

`rrinit.bin' now compensates for the ROM OS's stupid ^Z-stripping
code, which broke it in ZCN v1.1.

`cal', a calendar program like Unix `cal' - it's a Z80 port of a
public domain C program by Martin Minow.

`expr', a sort-of-clone of Unix `expr' - a Z80 port of a PD C program
by Erik Baalbergen.

Keyboard-reading code should now fully work around the bugs in the
NC100's keyboard hardware (see top of src/keyread.z for full details
of what the bugs are).

ZCN now uses IM2 by default, and can no longer be made to use IM1. So
the `useim2' command has been removed.

`setbaud' replaced with `stty'; e.g. `setbaud 7' translates to the
slightly more meaningful `stty 19200'.

New `cd -' (or `user -') command, go to last user area selected.

`note', a simple type-text-into-file internal command.

Zcnclock is now much nicer, and there's also `dclock', a new cut-down
version which fits in 1k.

`dir' now looks more like CP/M's dir.

Printer bug fixed - previously printing turned off the serial line
driver which, er, wasn't ideal. You'd never guess I hardly ever print
from ZCN, would you? :-)

Corrected serious bug in documentation. (The NC100's Z80 *DOES NOT*
run at 6MHz (it's more like 4.6MHz), even though it's a 6MHz-capable
chip! Also, ZCN isn't actually much faster than the ROM OS. It's still
better though. ;-))

Cursor move control code (^P) now uses 20h+y rather than the old (and
rather bizarre) 46h+y. 46h+y is still supported for now, but may not
be in future.

`ruler', displays a ruler onscreen.

`slide', a slide-rule emulator. Terribly useful. :-)

`calc' should now work on all ROM versions.

New `cboot' internal command cold-boots ZCN from the memory card.

New `capsctrl' internal command, swaps functions of Caps Lock and
Control keys if you do `capsctrl 1'.

New `crlf' internal command (not *another* internal command!) lets you
disable/re-enable the usual CP/M-ish line-break (CR/LF) after each
command finishes. Mainly intended for use in SUB files.

New `vdu' internal command (AIEEEE! :-)) prints one or two characters,
given their ASCII values. For example `vdu 1' clears the screen, and
`vdu 13 10' prints CR/LF.

You can now change the default LISTO setting for `bbcbas'. You have to
edit the executable, which isn't ideal, but zcn.txt demonstrates how
to do it with wade.

Some additions to zcnlib's 32-bit integer routines.

You can now change the address of the font used by ZCN. This lets

support/fish/fish.c
/* hack of BSD 'fish' */

/* Hitech C hasn't got strcasecmp().
 * I only bother doing an equality test, as that's all 'fish' uses it
 * for.
 */
int strcasecmp(char *s1,char *s2)
{
while(tolower(*s1)==tolower(*s2)) {
	if(*s1==0) return(0);	/* equal */
	s1++; s2++;
}

return(1);	/* different */
}

/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Muffy Barkocy.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1990 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)fish.c	5.4 (Berkeley) 1/18/91";
#endif /* not lint */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define	RANKS		13
#define	HANDSIZE	7
#define	CARDS		4

#define	USER		1
#define	COMPUTER	0
#define	OTHER(a)	(1 - (a))

char *cards[] = {
	"A", "2", "3", "4", "5", "6", "7",
	"8", "9", "10", "J", "Q", "K", NULL,
};
#define	PRC(card)	(void)printf(" %s", cards[card])

int promode;
int asked[RANKS], comphand[RANKS], deck[RANKS];
int userasked[RANKS], userhand[RANKS];


#define srandom srand
#define random rand


main(argc, argv)
	int argc;
	char **argv;
{
	unsigned char *ptr=(unsigned char *)255;
	int ch, move;

	if(argc==2 && strcmp(argv[1],"-p")==0)
		promode = 1;

/*	srandom(time((time_t *)NULL));*/
#asm
ld a,r
ld (255),a
#endasm
	srand(64*(*ptr)+*ptr);
	instructions();
	init();

	if (nrandom(2) == 1) {
		printplayer(COMPUTER);
		(void)printf("get to start.\n");
		goto istart;
	}
	printplayer(USER);
	(void)printf("get to start.\n");
	
	for (;;) {
		move = usermove();
		if (!comphand[move]) {
			if (gofish(move, USER, userhand))
				continue;
		} else {
			goodmove(USER, move, userhand, comphand);
			continue;
		}

istart:		for (;;) {
			move = compmove();
			if (!userhand[move]) {
				if (!gofish(move, COMPUTER, comphand))
					break;
			} else
				goodmove(COMPUTER, move, comphand, userhand);
		}
	}
	/* NOTREACHED */
}

usermove()
{
	register int n;
	register char **p;
	char buf[256];

	(void)printf("\nYour hand is:");
	printhand(userhand);

	for (;;) {
		(void)printf("You ask me for: ");
		(void)fflush(stdout);
		if (fgets(buf, BUFSIZ, stdin) == NULL)
			exit(0);
		if (buf[0] == '\0')
			continue;
		if (buf[0] == '\n') {
			(void)printf("%d cards in my hand, %d in the pool.\n",
			    countcards(comphand), countcards(deck));
			(void)printf("My books:");
			(void)countbooks(comphand);
			continue;
		}
		buf[strlen(buf) - 1] = '\0';
		if (!strcasecmp(buf, "p") && !promode) {
			promode = 1;
			(void)printf("Entering pro mode.\n");
			continue;
		}
		if (!strcasecmp(buf, "quit"))
			exit(0);
		for (p = cards; *p; ++p)
			if (!strcasecmp(*p, buf))
				break;
		if (!*p) {
			(void)printf("I don't understand!\n");
			continue;
		}
		n = p - cards;
		if (userhand[n]) {
			userasked[n] = 1;
			return(n);
		}
		if (nrandom(3) == 1)
			(void)printf("You don't have any of those!\n");
		else
			(void)printf("You don't have any %s's!\n", cards[n]);
		if (nrandom(4) == 1)
			(void)printf("No cheating!\n");
		(void)printf("Guess again.\n");
	}
	/* NOTREACHED */
}

compmove()
{
	static int lmove;

	if (promode)
		lmove = promove();
	else {
		do {
			lmove = (lmove + 1) % RANKS;
		} while (!comphand[lmove] || comphand[lmove] == CARDS);
	}
	asked[lmove] = 1;

	(void)printf("I ask you for: %s.  ", cards[lmove]);
	return(lmove);
}

promove()
{
	register int i, max;

	for (i = 0; i < RANKS; ++i)
		if (userasked[i] &&
		    comphand[i] > 0 && comphand[i] < CARDS) {
			userasked[i] = 0;
			return(i);
		}
	if (nrandom(3) == 1) {
		for (i = 0;; ++i)
			if (comphand[i] && comphand[i] != CARDS) {
				max = i;
				break;
			}
		while (++i < RANKS) 
			if (comphand[i] != CARDS &&
			    comphand[i] > comphand[max])
				max = i;
		return(max);
	} 
	if (nrandom(1024) == 0723) {
		for (i = 0; i < RANKS; ++i)
			if (userhand[i] && comphand[i])
				return(i);
	}
	for (;;) {
		for (i = 0; i < RANKS; ++i)
			if (comphand[i] && comphand[i] != CARDS &&
			    !asked[i])
				return(i);
		for (i = 0; i < RANKS; ++i)
			asked[i] = 0;
	}
	/* NOTREACHED */
}

drawcard(player, hand)
	int player;
	int *hand;
{
	int card;

	while (deck[card = nrandom(RANKS)] == 0);
	++hand[card];
	--deck[card];
	if (player == USER || hand[card] == CARDS) {
		printplayer(player);
		(void)printf("drew %s", cards[card]);
		if (hand[card] == CARDS) {
			(void)printf(" and made a book of %s's!\n",
			     cards[card]);
			chkwinner(player, hand);
		} else
			(void)printf(".  ");
	}
	return(card);
}

gofish(askedfor, player, hand)
	int askedfor, player;
	int *hand;
{
	printplayer(OTHER(player));
	(void)printf("say \"GO FISH!\"\n");
	if (askedfor == drawcard(player, hand)) {
		printplayer(player);
		(void)printf("drew the guess!  ");
		printplayer(player);
		(void)printf("get to ask again!\n");
		return(1);
	}
	return(0);
}

goodmove(player, move, hand, opphand)
	int player, move;
	int *hand, *opphand;
{
	printplayer(OTHER(player));
	(void)printf("have %d %s%s.  ",
	    opphand[move], cards[move], opphand[move] == 1 ? "": "'s");

	hand[move] += opphand[move];
	opphand[move] = 0;

	if (hand[move] == CARDS) {
		printplayer(player);
		(void)printf("made a book of %s's!\n", cards[move]);
		chkwinner(player, hand);
	}

	chkwinner(OTHER(player), opphand);

	printplayer(player);
	(void)printf("get another guess!\n");
}

chkwinner(player, hand)
	int player;
	register int *hand;
{
	register int cb, i, ub;

	for (i = 0; i < RANKS; ++i)
		if (hand[i] > 0 && hand[i] < CARDS)
			return;
	printplayer(player);
	(void)printf("don't have any more cards!\n");
	(void)printf("My books:");
	cb = countbooks(comphand);
	(void)printf("Your books:");
	ub = countbooks(userhand);
	(void)printf("\nI have %d, you have %d.\n", cb, ub);
	if (ub > cb) {
		(void)printf("\nYou win!!!\n");
		if (nrandom(1024) == 0723)
			(void)printf("Cheater, cheater, pumpkin eater!\n");
	} else if (cb > ub) {
		(void)printf("\nI win!!!\n");
		if (nrandom(1024) == 0723)
			(void)printf("Hah!  Stupid peasant!\n");
	} else
		(void)printf("\nTie!\n");
	exit(0);
}

printplayer(player)
	int player;
{
	switch (player) {
	case COMPUTER:
		(void)printf("I ");
		break;
	case USER:
		(void)printf("You ");
		break;
	}
}

printhand(hand)
	int *hand;
{
	register int book, i, j;

	for (book = i = 0; i < RANKS; i++)
		if (hand[i] < CARDS)
			for (j = hand[i]; --j >= 0;) 
				PRC(i);
		else
			++book;
	if (book) {
		(void)printf(" + Book%s of", book > 1 ? "s" : "");
		for (i = 0; i < RANKS; i++)
			if (hand[i] == CARDS)
				PRC(i);
	}
	(void)putchar('\n');
}

countcards(hand)
	register int *hand;
{
	register int i, count;

	for (count = i = 0; i < RANKS; i++)
		count += *hand++;
	return(count);
}

countbooks(hand)
	int *hand;
{
	int i, count;

	for (count = i = 0; i < RANKS; i++)
		if (hand[i] == CARDS) {
			++count;
			PRC(i);
		}
	if (!count)
		(void)printf(" none");
	(void)putchar('\n');
	return(count);
}

init()
{
	register int i, rank;

	for (i = 0; i < RANKS; ++i)
		deck[i] = CARDS;
	for (i = 0; i < HANDSIZE; ++i) {
		while (!deck[rank = nrandom(RANKS)]);
		++userhand[rank];
		--deck[rank];
	}
	for (i = 0; i < HANDSIZE; ++i) {
		while (!deck[rank = nrandom(RANKS)]);
		++comphand[rank];
		--deck[rank];
	}
}

nrandom(n)
	int n;
{
	int rand();

	return((int)rand() % n);
}

instructions()
{
#if 0
	int input;
	char buf[1024];

	(void)printf("Would you like instructions (y or n)? ");
	input = getchar();
	while (getchar() != '\n');
	if (input != 'y')
		return;

	(void)sprintf(buf, "%s %s", _PATH_MORE, _PATH_INSTR);
	(void)system(buf);
	(void)printf("Hit return to continue...\n");
	while ((input = getchar()) != EOF && input != '\n');
#endif
}

usage()
{
	(void)fprintf(stderr, "usage: fish [-p]\n");
	exit(1);
}
